src/herder/test/UpgradesTests.cpp

```diff
@@ -1029,6 +1029,301 @@ TEST_CASE("SCP timing config affects consensus behavior", "[upgrades][herder]")
+                cfg.mMemCostParams[ContractCostType::ParseWasmInstructions]
+                    .linearTerm *= 2;
+            });
+        ensureInMemorySizeIncreased();
+        // There is a small expected state size diff due to the settings upgrade
+        // contract.
+        verifyExpectedInMemorySize(100'000);
+
+        // The state size window must be unchanged.
+        REQUIRE(getStateSizeWindow() == p22StateSizeWindow);
+    }
+
+    INFO("upgrade to p23");
+    executeUpgrade(test.getApp(), makeProtocolVersionUpgrade(23));
+    // In-memory size shouldn't have changed as it has been computed with p23
+    // logic.
+    REQUIRE(test.getApp().getLedgerManager().getSorobanInMemoryStateSize() ==
+            lastInMemorySize);
+    auto const p23MemorySize = lastInMemorySize;
+    // State size window now contains only the current in-memory size.
+    expectSingleValueStateSizeWindow(p23MemorySize);
+    verifyAverageStateSize(p23MemorySize, p23MemorySize);
+
+    {
+        INFO("fill window with in-memory size in p23");
+        closeLedger(test.getApp());
+
+        // Now generate more txs to fill up the window with in-memory sizes.
+        generateTxs(windowSize * samplePeriod * 4);
+        verifyExpectedInMemorySize(0);
+        REQUIRE(getStateSizeWindow().back() == lastInMemorySize);
+        verifyAverageStateSize(p23MemorySize, lastInMemorySize);
+    }
+
+    {
+        INFO("upgrade memory settings in p23 without state size snapshot");
+        // Make sure we won't snapshot the window size when we perform the
+        // upgrade on LCL + 1.
+        while (test.getLCLSeq() % windowSize == windowSize - 2)
+        {
+            closeLedger(test.getApp());
+        }
+
+        modifySorobanNetworkConfig(
+            test.getApp(), [](SorobanNetworkConfig& cfg) {
+                cfg.mMemCostParams[ContractCostType::ParseWasmInstructions]
+                    .constTerm *= 3;
+                cfg.mMemCostParams[ContractCostType::ParseWasmInstructions]
+                    .linearTerm *= 3;
+            });
+        ensureInMemorySizeIncreased();
+        verifyExpectedInMemorySize(100'000);
+        expectSingleValueStateSizeWindow(lastInMemorySize);
+    }
+
+    {
+        INFO("upgrade memory settings in p23 with state size snapshot");
+        // Wait until we're one ledger before the ledger that will trigger
+        // snapshotting.
+        while (test.getLCLSeq() % windowSize == windowSize - 1)
+        {
+            closeLedger(test.getApp());
+        }
+
+        modifySorobanNetworkConfig(
+            test.getApp(), [](SorobanNetworkConfig& cfg) {
+                cfg.mMemCostParams[ContractCostType::ParseWasmInstructions]
+                    .constTerm *= 2;
+                cfg.mMemCostParams[ContractCostType::ParseWasmInstructions]
+                    .linearTerm *= 2;
+            });
+        ensureInMemorySizeIncreased();
+        verifyExpectedInMemorySize(200'000);
```

Comment:
I assume this increased to 200'000 because we have 100'00 from the first upgrade, and another 100'000 from this upgrade, right?

Reply:
No, we keep correcting for the last diff observed. But since we increase the costs exponentially, the discrepancy magnitude goes up as well.